### PR TITLE
JAMES-3762 Reworks MailRepository contract to use MailKeys instead of Mails

### PR DESCRIPTION
**Pull request by mbaechler**
Co-Authored-By: Matthieu Baechler <matthieu@apache.org>

### CODE REVIEW

File: server/mailrepository/mailrepository-api/src/test/java/org/apache/james/mailrepository/MailRepositoryContract.java [hunk 1]
```diff
@@ -206,19 +206,19 @@ default void shouldPreserveDsnParameters() throws Exception {
...
 
-        assertThat(testee.retrieve(MAIL_1).dsnParameters()).contains(dsnParameters);
+        assertThat(testee.retrieve(key).dsnParameters()).contains(dsnParameters);
     }
 
     @Test
     default void retrieveShouldReturnNullAfterRemoveAll() throws Exception {
         MailRepository testee = retrieveRepository();
-        testee.store(createMail(MAIL_1));
+        var key = testee.store(createMail(MAIL_1));
```
Comment (chibenwa) [inline server/mailrepository/mailrepository-api/src/test/java/org/apache/james/mailrepository/MailRepositoryContract.java:217]: I thought we once upon a time decided to not use `var` as it looses type explicitness.

File: server/protocols/webadmin/webadmin-mailrepository/src/main/java/org/apache/james/webadmin/service/ReprocessingService.java [hunk 2]
```diff
@@ -86,12 +86,12 @@ static class Reprocessor implements Closeable {
             this.configuration = configuration;
         }
 
-        private void reprocess(MailRepository repository, Mail mail) {
+        private void reprocess(MailRepository repository, Mail mail, MailKey key) {
```
Comment (chibenwa) [inline server/protocols/webadmin/webadmin-mailrepository/src/main/java/org/apache/james/webadmin/service/ReprocessingService.java:89]: If we have the mail we have the key? IE this method signature can likely stay the same.

**Review by chibenwa — APPROVED** [review]

**Review by Arsnael — APPROVED** [review]

**Comment by Arsnael** [comment]
```
22:20:52,265 [INFO] There is 1 error reported by Checkstyle 8.29 with checkstyle.xml ruleset.

22:20:52,265 [ERROR] src/main/java/org/apache/james/mailrepository/jdbc/JDBCMailRepository.java:[37,8] (imports) UnusedImports: Unused import - java.util.Collection.
```

**Comment by jeantil** [comment]
I have pushed a fix for the checkstyle violation, we will rebase -i and add the ticket number in the commit title once the CI clears. I didn't want to force push twice

**Comment by Arsnael** [comment]
```
07:21:12,685 [ERROR] Failed to execute goal org.apache.maven.plugins:maven-checkstyle-plugin:3.1.0:check (check-style) on project james-server-mailrepository-memory: You have 1 Checkstyle violation. -> [Help 1]

org.apache.maven.lifecycle.LifecycleExecutionException: Failed to execute goal org.apache.maven.plugins:maven-checkstyle-plugin:3.1.0:check (check-style) on project james-server-mailrepository-memory: You have 1 Checkstyle violation.
```
An other one... Couldn't see it in the logs this time though, might want to do some digging :)

**Comment by jeantil** [comment]
local build was green, I wonder if checkstyle could be moved from compile to validate: compiling james is pretty heavy ...

**Comment by mbaechler** [comment]
squash and add ticket number then push force

**Pull request merged**